### ISSUE TITLE
trailing-comma: esSpecCompliant not applied to function call

### ISSUE THREAD

**Issue by jmlopez-rod** [palantir/tslint#3870]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.7.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
// good (note that a comma must not appear after a "rest" element)
function createHero(
  firstName,
  lastName,
  inventorOf,
  ...heroArgs
) {
  // does nothing
}

// bad
createHero(
  firstName,
  lastName,
  inventorOf
);

// good
createHero(
  firstName,
  lastName,
  inventorOf,
);

// good (note that a comma must not appear after a "rest" element)
createHero(
  firstName,
  lastName,
  inventorOf,
  ...heroArgs
);
```

with `tslint.json` configuration:

```json
    "trailing-comma": [
      true,
      {
        "multiline": {
          "arrays": "always",
          "exports": "always",
          "functions": "always",
          "imports": "always",
          "objects": "always",
          "typeLiterals": "always"
        },
        "singleline": {
          "arrays": "never",
          "exports": "never",
          "functions": "never",
          "imports": "never",
          "objects": "never",
          "typeLiterals": "never"
        },
        "esSpecCompliant": true
      }
    ],
```

#### Actual behavior

```
tslint testFile.ts 

ERROR: testFile.ts[15, 13]: Missing trailing comma
ERROR: testFile.ts[30, 14]: Missing trailing comma
```

#### Expected behavior

Should only report the missing trailing comma on line 15. The `esSpecCompliant` seems to do its job when we define a function but not when we use it.

**Comment by peterjwest** [comment]
I may have missed something here, but I don't think this makes sense.

When you define a function with a spread operator, this must be the last argument due to the pattern matching, so a comma doesn't make sense.
```ts
function createHero(
  firstName,
  lastName,
  inventorOf,
  ...heroArgs
) {}
```

However when you call a function, you can add as many spread operators as you want because we are talking about a concrete array with a known length. So a trailing comma makes as much sense as it does anywhere:
```ts
createHero(
  'Foo',
  'Bar',
  'zim',
  ...[1, 2, 3],
  ...[4, 5, 6],
) {}
```

**Comment by jmlopez-rod** [comment]
This is about the `esSpecCompliant` option.

> To align this rule with the ECMAScript specification that is implemented in modern JavaScript VMs, ? > there is a third option esSpecCompliant. Set this option to true to disallow trailing comma on object
> and array rest and rest parameters.

**Comment by adidahiya** [comment]
@jmlopez-rod according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas#Illegal_trailing_commas, trailing commas in function call argument lists are _not_ illegal, so I'm not sure there needs to be a behavior change here. Can you point me to any additional documentation suggesting that we need to change the rule?

**Comment by jmlopez-rod** [comment]
@adidahiya , I'm not arguing that its bad. Its the documentation that states:

>  Set this option to true to disallow trailing comma on object and array rest and rest parameters.

So looking back at my original post

> Should only report the missing trailing comma on line 15.

Which it does. Its complaining at this code:

```
// bad
createHero(
  firstName,
  lastName,
  inventorOf
);
```

We need to add the trailing comma after `inventorOf`, that is in line 15.

Now, the last one

```
// good (note that a comma must not appear after a "rest" element)
createHero(
  firstName,
  lastName,
  inventorOf,
  ...heroArgs
);
```

It complains that it needs a comma after `...heroArgs` (It should not do that). Since I have the esSpecCompliant flag it should not report that we need a trailing comma since `heroArgs` is a rest parameter.

Again, the docs for the flag states:

> disallow trailing comma on object and array rest and rest parameters.

In other words, if I were to write

```
// good (note that a comma must not appear after a "rest" element)
createHero(
  firstName,
  lastName,
  inventorOf,
  ...heroArgs,
);
```

Then it should complain saying something along the lines of `esSpecComplaint flag disallows the use of trailing commas after rest parameters`.

**Comment by jmlopez-rod** [comment]
Since this is ok:

<img width="849" alt="Screen Shot 2019-06-19 at 10 07 17 PM" src="https://user-images.githubusercontent.com/1810397/59815594-a7f8b480-92de-11e9-8890-6af7c3d0e3a2.png">

It seems that I missed one important detail here:

A rest parameter refers to the last parameter in the declaration of a function. This is not be the same as the spread arguments (in a function call).

I think I may have been confused by the flag when I first saw it because typescript already complains if you try to use the trailing comma, because of this it made me ask, what is the use of the flag?

<img width="400" alt="Screen Shot 2019-06-19 at 10 24 36 PM" src="https://user-images.githubusercontent.com/1810397/59816294-2ce4cd80-92e1-11e9-82a9-68a2d0144764.png">

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
